include/roaring/portability.h

```diff
@@ -399,41 +399,130 @@ static inline int roaring_hamming(uint64_t x) {
  #endif // __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
 #endif
 
-#if defined(__cplusplus) && __cplusplus >= 201103L
-#ifdef __has_include
-#if __has_include(<atomic>)
-// It is now safe:
-#define CROARING_CPP_ATOMIC 1
-#define CROARING_C_ATOMIC 0
-#define CROARING_CPP_WINDOWS_ATOMIC 0
+// Defines for the possible CROARING atomic implementations
+#define CROARING_ATOMIC_IMPL_NONE          1
+#define CROARING_ATOMIC_IMPL_CPP           2
+#define CROARING_ATOMIC_IMPL_C             3
+#define CROARING_ATOMIC_IMPL_C_WINDOWS     4
+
+// If the use has forced a specific implementation, use that, otherwise,
+// figure out the best implementation we can use.
+#if !defined(CROARING_ATOMIC_IMPL)
+  #if defined(__cplusplus) && __cplusplus >= 201103L
+    #ifdef __has_include
+      #if __has_include(<atomic>)
+        #define CROARING_ATOMIC_IMPL CROARING_ATOMIC_IMPL_CPP
+      #endif //__has_include(<atomic>)
+    #else
+       // We lack __has_include to check:
+       #define CROARING_ATOMIC_IMPL CROARING_ATOMIC_IMPL_CPP
+    #endif //__has_include
+  #elif !defined(__STDC_NO_ATOMICS__)
+    #define CROARING_ATOMIC_IMPL CROARING_ATOMIC_IMPL_C
+  #else // (c, no atomics)
+    #if CROARING_REGULAR_VISUAL_STUDIO
```

Comment:
I think here you have reversed the logic. You need to check for visual studio first, not for __STDC_NO_ATOMICS__

Reply:
That is, if CROARING_REGULAR_VISUAL_STUDIO is true, then you must not pick CROARING_ATOMIC_IMPL_C, at all.

I know that it does not make sense, but that's how I had to do it.

Look at my code...

```C
if CROARING_REGULAR_VISUAL_STUDIO
 // https://www.technetworkhub.com/c11-atomics-in-visual-studio-2022-version-17/
 #define CROARING_ATOMIC 0
 #define CROARING_CPP_ATOMIC 0
 #define CROARING_CPP_WINDOWS_ATOMIC 1
 # include <intrin.h>
 # pragma intrinsic (_InterlockedIncrement)
 # pragma intrinsic (_InterlockedDecrement)
 #elif defined(__STDC_NO_ATOMICS__)
 #define CROARING_ATOMIC 0
 #define CROARING_CPP_ATOMIC 0
 #define CROARING_CPP_WINDOWS_ATOMIC 0
 #else // C
 #define CROARING_C_ATOMIC 1
 #define CROARING_CPP_ATOMIC 0
 #define CROARING_CPP_WINDOWS_ATOMIC 0
 #include <stdatomic.h>
 #endif
```

It is organized in this way for a reason: that's what we need to do.

Reply:
I'm trying another method. I would really like to automatically pick up c atomics on windows if they work, rather than using interlocked stuff.

I think the important part is that `__STDC_NO_ATOMICS__` _can_ only be set if we're at least c11, if we're not reported as at least c11, `__STDC_NO_ATOMICS__` will _also_ be unset, because it the idea of it doesn't exist yet. So I changed it to check first for `__STDC_VERSION__ >= c11_version && !defined(__STDC_NO_ATOMICS__)`, and then MSVC.